_posts/2023-01-07-datafusion-16.0.0.md

```diff
@@ -0,0 +1,289 @@
+* Sort and partition aware optimizations such as  #3969 and  #4691,  skipping potentially expensive operations
+* Basic filter selectivity analysis (#3868)
+
+
+In the coming few months, we plan work on:
+* Improved grouping performance (TODO link)
+* bloom filtering
+* investigate RLE (Run End Encoding support) (todo Arrow link)
+* Enable predicate pushdown by default for all cases
+* OTHERS?
+
+## Runtime Resource Limits
+
+Initially, DataFusion could potentially use unbounded amounts of memory for certain queries that included Sorts, Grouping or Joins.
+
+In version 16.0.0, it is possible to limit DataFusion's memory usage for Sorting and Grouping. We are looking for help adding similar limiting for Joins as well as expanding our algorithms to spill to secondary storage, if available. See #3941 fore more detail.
+
+
+## SQL Window Function
+[SQL window functions](https://en.wikipedia.org/wiki/Window_function_(SQL))  are useful for a variety of analysis and DataFusion's implementation is close to complete now.
+
+- Custom window frames such as `... OVER (ORDER BY ... RANGE BETWEEN 0.2 PRECEDING AND 0.2 FOLLOWING)`
+- Unbounded window frames such as `... OVER (ORDER BY ... RANGE UNBOUNDED ROWS PRECEDING)`
+- Support for the `NTILE` window function (#4676)
+- Support for `GROUPS` mode (#4155)
+
+
+# Improved Join Support
+
+Joins are often the most complicated operations to handle well in
+analytics systems and DataFusion 16.0.0 offers significant improvements
+such as
+
+- Cost based optimizer (CBO) that can automatically reorder join evaluations, algorithm (Merge / Hash), and pick build side based on available statistics join type (`INNER`, `LEFT`, etc) (#4219)
+- Fast non `column=column` equijoins such as `JOIN ON a.x + 5 = b.y`
+- Better optimized non-equijoins (nested loops rather than cross join and filter) (#4562) <!-- TODO is this a good thing to mention as any time this is usd the query is going to go slow or the data size is small -->
+
+# Streaming Execution
+
+TODO: write up some supporting text introducing this concept and explaining that DataFusion can now be used for stream processing, much like [Spark Streaming](https://www.databricks.com/glossary/what-is-spark-streaming)
+
+Support for executing infinite files and boundedness-aware join reordering rule (#4694)
+
+# Better Support for Distributed catalogs
+
+With the rise of distributed catalog / metadata stores such as
+delta.io and iceberg (TODO get links to rust projects), it is
+increasingly important to be enable asynchronous I/O to those remote
+catalogs during query planning rather than requiring up front fast
+local access to all relevant catalog information.
+
+
+DataFusion has been enhanced with better support for such asynchronous catalogs (#4607) , which also resulted in improved configuration management APIs.
+
+
+
+# Additional SQL support
+SQL support continues to improve, including some of these highlights:
```

Comment:
are there other things to highlight? @Ted-Jiang / @liukun4515 ?

Reply:
@alamb I can not find any more